Neos.Neos/Classes/Controller/Frontend/ContentSubgraphCacheWarmup.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\Neos\Controller\Frontend;
+
+use Neos\ContentRepository\Core\Projection\ContentGraph\ContentSubgraphInterface;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Filter\FindSubtreeFilter;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Node;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Nodes;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Subtree;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\ContentRepositoryRegistry\SubgraphCachingInMemory\ContentSubgraphWithRuntimeCaches;
+use Neos\ContentRepositoryRegistry\SubgraphCachingInMemory\SubgraphCachePool;
+use Neos\Flow\Annotations as Flow;
+use Neos\Neos\Domain\Service\NodeTypeNameFactory;
+
+/**
+ * Cache warmup for rendering
+ *
+ * When rendering a document its child content nodes are regularly traversed.
+ * This is done conventionally via fusion and flow query traversing the children
+ * via {@see ContentSubgraphInterface::findChildNodes()} recursively.
+ *
+ * To optimise rendering we fetch the subtree, excluding nested document nodes, effectively all child content nodes
+ * of the entry document and use those to fill the cache.
+ *
+ * By ensuring we use {@see ContentRepositoryRegistry::subgraphForNode()} in flow queries, we access the cached subgraph
+ * via {@see ContentSubgraphWithRuntimeCaches}
+ *
+ * Reference loading of nodes is not optimised.
+ *
+ * TODO: Add tests that caches are filled correctly
+ * TODO: Add performance tests as that the findSubtree() CTE is indeed faster, than multiple child node queries.
```

Comment:
the first iteration of the warmup was already part of the FIRST new node controller commit: https://github.com/neos/neos-development-collection/commit/2c647b839c0bdeaa33bf5dc87013694c5b34f04c

This means the logic is well tested over time, but also that there is probably no data on how this improves actual uncached renderings.
Especially as we know that recursive CTE's are no wonder weapon. It might be equally fast to have `findChildNodes` build up the caches as we go.

As one thing is sure: in case a page is fully cached by fusion, this will be actually counter productive to fetch all child nodes.

Reply:
yep, indeed...